File: manager_app/src/test/java/com/expense/manager/e2e/steps/LoginSteps.java
package com.expense.manager.e2e.steps;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import com.expense.manager.e2e.hooks.Hooks;
import java.time.Duration;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import com.expense.manager.e2e.pages.LoginPage;

import io.cucumber.java.After;
import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class LoginSteps {
    private WebDriver driver;
    private static final String LOGIN_URL = "http://localhost:5173/";
    private LoginPage loginPage;

    @Before(order = 1)
    public void setUpPages() {
        driver = Hooks.driver;
        loginPage = new LoginPage(driver);
    }

    @After(order = 1)
    public void tearDown() {
        // Driver is cleaned up by Hooks
    }

    @Given("the user is on the manager login page")
    public void the_user_is_on_the_manager_login_page() {
        driver.get(LOGIN_URL);

        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(2));

        WebElement usernameField = wait.until(
                ExpectedConditions.visibilityOfElementLocated(By.id("username"))
        );

        assertTrue(usernameField.isDisplayed());
    }
    @When("the user enters username {string}")
    public void the_user_enters_username(String username) {
        loginPage.enterUsername(username);
    }
    @When("the user enters password {string}")
    public void the_user_enters_password(String password) {
        loginPage.enterPassword(password);
    }
    @When("the user clicks the login button")
    public void the_user_clicks_the_login_button() {
        loginPage.clickLogin();
    }

    @Then("the login result should be {string}")
    public void the_login_result_should_be(String result) {
        if (result.equals("success")) {
            new WebDriverWait(driver, Duration.ofSeconds(10))
                .until(ExpectedConditions.urlContains("/manager"));

            assertTrue(driver.getCurrentUrl().contains("/manager"));
        } else {
            WebElement error = new WebDriverWait(driver, Duration.ofSeconds(10))
                .until(ExpectedConditions.visibilityOfElementLocated(
                    By.cssSelector("p.text-red-600")));

            assertEquals("Username or password not valid.", error.getText());

            // Optional: verify failed login stays on login page
            assertTrue(driver.getCurrentUrl().equals(LOGIN_URL));
        }
    }
}

File: employee_app/e2e/features/steps/login_steps.py
from behave import given, when, then
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


@given("the app is launched")
def step_launch_app(context):
    context.driver.get(context.base_url)


@when('I enter username "{username}" and password "{password}"')
def step_enter_credentials(context, username, password):
    context.driver.find_element(By.ID, "username").send_keys(username)
    context.driver.find_element(By.ID, "password").send_keys(password)


@when("I click the login button")
def step_click_login(context):
    context.driver.find_element(
        By.CSS_SELECTOR, "button[type='submit']"
    ).click()


@then("the employee dashboard is displayed")
def step_dashboard_displayed(context):
    WebDriverWait(context.driver, 10).until(
        EC.text_to_be_present_in_element(
            (By.TAG_NAME, "h1"), "Employee Expense Portal"
        )
    )


@then("an error message is displayed")
def step_error_displayed(context):
    WebDriverWait(context.driver, 10).until(
        EC.text_to_be_present_in_element(
            (By.TAG_NAME, "form"), "Username or password not valid."
        )
    )
